File: src/speedometer_py/speedometer_py/speedometer_node.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import PoseStamped, Twist
from collections import deque
import numpy as np
import math

class SpeedometerNode(Node):
    def __init__(self):
        super().__init__('speedometer')

        # 1. Setup Subscription and Publisher
        self.pose_sub = self.create_subscription(
            PoseStamped,
            'vicon_pose',
            self.cal_velocities,
            2
        )
        self.velocity_pub = self.create_publisher(Twist, 'velocity', 1)

        # 2. Declare and Get Parameters
        self.declare_parameter('median_window_size', 5)
        self.declare_parameter('median_threshold', 2.0)  # rad/s

        self.median_window_size = self.get_parameter('median_window_size').value
        self.median_threshold = self.get_parameter('median_threshold').value

        # 3. Initialize State Variables
        self.last_pose_msg = None
        # Using deque with maxlen automatically handles popping old values!
        self.angular_velo_history = deque(maxlen=self.median_window_size)
        self.VICON_DT = 0.01  # 100Hz from Vicon

        self.get_logger().info(
            f"Speedometer_node has started with median filter "
            f"(window={self.median_window_size}, threshold={self.median_threshold:.2f} rad/s)"
        )

    def apply_median_filter(self, new_value: float) -> float:
        """Applies a median filter to reject outlier angular velocity spikes."""
        self.angular_velo_history.append(new_value)

        # Need at least 3 values for meaningful median
        if len(self.angular_velo_history) < 3:
            return new_value

        median = float(np.median(self.angular_velo_history))

        # Check if new value is an outlier (beyond threshold from median)
        if abs(new_value - median) > self.median_threshold:
            self.get_logger().debug(
                f"Angular velocity {new_value:.3f} rad/s rejected "
                f"(median={median:.3f}, diff={abs(new_value - median):.3f} > threshold={self.median_threshold:.3f})"
            )
            return median

        return new_value

    def get_yaw_from_quaternion(self, q) -> float:
        """Helper to convert a quaternion message into a yaw angle (radians)."""
        siny_cosp = 2.0 * (q.w * q.z + q.x * q.y)
        cosy_cosp = 1.0 - 2.0 * (q.y * q.y + q.z * q.z)
        return math.atan2(siny_cosp, cosy_cosp)

    def cal_velocities(self, pose_msg):
        """Callback to calculate linear and angular velocities from pose data."""
        if self.last_pose_msg is None:
            self.last_pose_msg = pose_msg
            return

        # Calculate x and y linear velocities
        dx = pose_msg.pose.position.x - self.last_pose_msg.pose.position.x
        dy = pose_msg.pose.position.y - self.last_pose_msg.pose.position.y
        
        x_velo = dx / self.VICON_DT
        y_velo = dy / self.VICON_DT

        # Calculate angular velocity (yaw)
        current_yaw = self.get_yaw_from_quaternion(pose_msg.pose.orientation)
        last_yaw = self.get_yaw_from_quaternion(self.last_pose_msg.pose.orientation)

        delta_yaw = current_yaw - last_yaw

        # Assume the car took the shortest path to reach its current angle
        if delta_yaw > math.pi:
            delta_yaw -= 2 * math.pi
        elif delta_yaw < -math.pi:
            delta_yaw += 2 * math.pi

        angular_velo = delta_yaw / self.VICON_DT

        # Apply median filter
        filtered_angular_velo = self.apply_median_filter(angular_velo)

        # The vicon angle samples have jitter, round to tenths place (hundredths natively)
        filtered_angular_velo = round(filtered_angular_velo * 100) / 100.0

        # Publish the Velocity Message
        velocity_msg = Twist()
        velocity_msg.linear.x = float(x_velo)
        velocity_msg.linear.y = float(y_velo)
        velocity_msg.angular.z = float(filtered_angular_velo)

        self.velocity_pub.publish(velocity_msg)

        # Debug Logging
        raw_deg = round((angular_velo * (180.0 / math.pi)) * 100) / 100.0
        filt_deg = round((filtered_angular_velo * (180.0 / math.pi)) * 100) / 100.0
        
        self.get_logger().debug(
            f"X velocity: {x_velo:.3f} m/s, Y velocity: {y_velo:.3f} m/s, "
            f"Angular velocity: {filt_deg:.3f} degree/sec (raw: {raw_deg:.3f})"
        )

        self.last_pose_msg = pose_msg

def main(args=None):
    rclpy.init(args=args)
    node = SpeedometerNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

if __name__ == '__main__':
    main()

File: src/speedometer/src/speedometer_node.cpp
#include <tf2/LinearMath/Quaternion.h>

#include <algorithm>
#include <cmath>
#include <deque>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <geometry_msgs/msg/twist.hpp>
#include <memory>
#include <rclcpp/logging.hpp>
#include <rclcpp/rclcpp.hpp>
#include <tf2/LinearMath/Matrix3x3.hpp>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <vector>

namespace
{
using twist_msg = geometry_msgs::msg::Twist;
using twist_pub = rclcpp::Publisher<twist_msg>::SharedPtr;
using pose_msg = geometry_msgs::msg::PoseStamped;
using pose_msg_ptr = pose_msg::SharedPtr;
using pose_sub = rclcpp::Subscription<pose_msg>::SharedPtr;

using namespace std::placeholders;
using std::bind;
using std::make_shared;
}  // namespace

class Speedometer_Node : public rclcpp::Node
{
   public:
    Speedometer_Node() : Node("speedometer")
    {
        // Better QoS settings

        // get pose msg for vicon reader node
        pose_sub_ = this->create_subscription<pose_msg>(
            "vicon_pose", 2, bind(&Speedometer_Node::cal_velocities, this, _1));

        // velocity publisher
        velocity_pub_ = this->create_publisher<twist_msg>("velocity", 1);

        // Declare parameters for median filter
        this->declare_parameter("median_window_size", 5);
        this->declare_parameter("median_threshold", 2.0);  // rad/s

        median_window_size_ = this->get_parameter("median_window_size").as_int();
        median_threshold_ = this->get_parameter("median_threshold").as_double();

        RCLCPP_INFO(
            this->get_logger(),
            "Speedometer_node has started with median filter (window=%d, threshold=%.2f rad/s)",
            median_window_size_, median_threshold_);
    }

   private:
    pose_sub pose_sub_;
    twist_pub velocity_pub_;
    pose_msg_ptr last_pose_msg;

    // Median filter variables
    std::deque<double> angular_velo_history_;
    int median_window_size_;
    double median_threshold_;

    // the vicon record data at 100hz
    const double VICON_DT = 0.01;  // therefore 0.01 sec period

    double apply_median_filter(double new_value)
    {
        // Add new value to history
        angular_velo_history_.push_back(new_value);

        // Maintain window size
        if (angular_velo_history_.size() > static_cast<size_t>(median_window_size_))
        {
            angular_velo_history_.pop_front();
        }

        // Need at least 3 values for meaningful median
        if (angular_velo_history_.size() < 3)
        {
            return new_value;
        }

        // Calculate median
        std::vector<double> sorted_values(angular_velo_history_.begin(),
                                          angular_velo_history_.end());
        std::sort(sorted_values.begin(), sorted_values.end());

        double median;
        size_t n = sorted_values.size();
        if (n % 2 == 0)
        {
            median = (sorted_values[n / 2 - 1] + sorted_values[n / 2]) / 2.0;
        }
        else
        {
            median = sorted_values[n / 2];
        }

        // Check if new value is within threshold of median
        if (std::abs(new_value - median) > median_threshold_)
        {
            RCLCPP_DEBUG(
                this->get_logger(),
                "Angular velocity %.3f rad/s rejected (median=%.3f, diff=%.3f > threshold=%.3f)",
                new_value, median, std::abs(new_value - median), median_threshold_);
            return median;
        }

        return new_value;
    }

    void cal_velocities(pose_msg_ptr pose_msg)
    {
        double x_velo;        // x velocity
        double y_velo;        // y velocity
        double angular_velo;  // how fast the car is turning left and right
        tf2::Quaternion last_quat;
        tf2::Quaternion current_quat;
        double last_yaw;
        double current_yaw;
        double delta_yaw;
        twist_msg velocity_msg;

        // check if the last_pose_msg is empty
        if (!last_pose_msg)
        {
            last_pose_msg = pose_msg;
            return;  // end early
        }

        // calculate x, y, and z velocities
        x_velo = (pose_msg->pose.position.x - last_pose_msg->pose.position.x) / VICON_DT;
        y_velo = (pose_msg->pose.position.y - last_pose_msg->pose.position.y) / VICON_DT;

        tf2::fromMsg(pose_msg->pose.orientation, current_quat);
        tf2::fromMsg(last_pose_msg->pose.orientation, last_quat);

        // since it's a car I only care about the yaw
        double dont_care;
        tf2::Matrix3x3(current_quat).getRPY(dont_care, dont_care, current_yaw);
        tf2::Matrix3x3(last_quat).getRPY(dont_care, dont_care, last_yaw);

        // get difference
        delta_yaw = current_yaw - last_yaw;

        // we assume the car took the shortest path
        // to reach it current angle.
        if (delta_yaw > M_PI) delta_yaw -= 2 * M_PI;
        if (delta_yaw < -M_PI) delta_yaw += 2 * M_PI;

        angular_velo = delta_yaw / VICON_DT;

        // Apply median filter to angular velocity
        double filtered_angular_velo = apply_median_filter(angular_velo);

        // the vicon angle samples has jitter
        // so round to tenths place
        filtered_angular_velo = round(100 * filtered_angular_velo) / 100;

        // velocity message
        velocity_msg.linear.x = x_velo;
        velocity_msg.linear.y = y_velo;
        velocity_msg.angular.z = filtered_angular_velo;

        velocity_pub_->publish(velocity_msg);

        // update last pose
        last_pose_msg = pose_msg;

        // debug logging
        RCLCPP_DEBUG(this->get_logger(),
                     "X velocity: %.3f m/s, Y velocity: %.3f m/s, Angular velocity: %.3f "
                     "degree/sec (raw: %.3f)",
                     x_velo, y_velo, round(100 * filtered_angular_velo * (180 / M_PI)) / 100,
                     round(100 * angular_velo * (180 / M_PI) / 100));

        RCLCPP_DEBUG(this->get_logger(),
                     "current and last: X: %.3fm, %.3fm, Y: %.3fm, %.3fm, Yaw: %.3f°, %.3f°",
                     pose_msg->pose.position.x, last_pose_msg->pose.position.x,
                     pose_msg->pose.position.y, last_pose_msg->pose.position.y,
                     current_yaw * (180 / M_PI), last_yaw * (180 / M_PI));
    }
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = make_shared<Speedometer_Node>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}